presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/AddExchangesBelowPartialAggregationOverGroupIdRuleSet.java

```diff
@@ -0,0 +1,349 @@
+
+    private static final Pattern<ExchangeNode> WITHOUT_PROJECTION =
+            // If there was no exchange here, adding new exchanges could break property derivations logic of AddExchanges, AddLocalExchanges
+            typeOf(ExchangeNode.class)
+                    .with(scope().equalTo(REMOTE))
+                    .with(source().matching(
+                            typeOf(AggregationNode.class).capturedAs(AGGREGATION)
+                                    .with(step().equalTo(AggregationNode.Step.PARTIAL))
+                                    .with(nonEmpty(groupingColumns()))
+                                    .with(source().matching(
+                                            typeOf(GroupIdNode.class).capturedAs(GROUP_ID)))));
+
+    private static final double GROUPING_SETS_SYMBOL_REQUIRED_FREQUENCY = 0.5;
+    private static final double ANTI_SKEWNESS_MARGIN = 3;
+
+    private final Metadata metadata;
+    private final SqlParser parser;
+    private final TaskCountEstimator taskCountEstimator;
+    private final DataSize maxPartialAggregationMemoryUsage;
+
+    public AddExchangesBelowPartialAggregationOverGroupIdRuleSet(
+            Metadata metadata,
+            SqlParser parser,
+            TaskCountEstimator taskCountEstimator,
+            TaskManagerConfig taskManagerConfig)
+    {
+        this.metadata = requireNonNull(metadata, "metadata is null");
+        this.parser = requireNonNull(parser, "parser is null");
+        this.taskCountEstimator = requireNonNull(taskCountEstimator, "taskCountEstimator is null");
+        this.maxPartialAggregationMemoryUsage = requireNonNull(taskManagerConfig, "taskManagerConfig is null").getMaxPartialAggregationMemoryUsage();
+    }
+
+    public Set<Rule<?>> rules()
+    {
+        return ImmutableSet.of(
+                new AddExchangesBelowProjectionPartialAggregationGroupId(),
+                new AddExchangesBelowExchangePartialAggregationGroupId());
+    }
+
+    private class AddExchangesBelowProjectionPartialAggregationGroupId
+            extends BaseAddExchangesBelowExchangePartialAggregationGroupId
+    {
+        @Override
+        public Pattern<ExchangeNode> getPattern()
+        {
+            return WITH_PROJECTION;
+        }
+
+        @Override
+        public Result apply(ExchangeNode exchange, Captures captures, Context context)
+        {
+            ProjectNode project = captures.get(PROJECTION);
+            AggregationNode aggregation = captures.get(AGGREGATION);
+            GroupIdNode groupId = captures.get(GROUP_ID);
+            return transform(aggregation, groupId, context)
+                    .map(newAggregation -> {
+                        PlanNode newProject = project.replaceChildren(ImmutableList.of(newAggregation));
+                        PlanNode newExchange = exchange.replaceChildren(ImmutableList.of(newProject));
+                        return Result.ofPlanNode(newExchange);
```

Comment:
I would inline these, then structure of the code matches the structure of the plan.